plugins/inputs/vsphere/vsan.go

```diff
@@ -0,0 +1,523 @@
+	if versionLowerThan(e.apiVersion, "6.7") {
+		log.Printf("I! [inputs.vsphere][vSAN] Minimum API Version 6.7 required for resync summary. Found: %s. Skipping VCenter: %s", e.apiVersion, e.URL.Host)
+		return nil
+	}
+	hosts, err := clusterObj.Hosts(ctx)
+	if err != nil {
+		return err
+	}
+	if len(hosts) == 0 {
+		log.Printf("I! [inputs.vsphere][vSAN]\tNo host in cluster: %s", clusterObj.Name())
+		return nil
+	}
+	hostRefValue := hosts[0].Reference().Value
+	vsanSystemEx := types.ManagedObjectReference{
+		Type:  "VsanSystemEx",
+		Value: fmt.Sprintf("vsanSystemEx-%s", strings.Split(hostRefValue, "-")[1]),
+	}
+
+	includeSummary := true
+	request := vsantypes.VsanQuerySyncingVsanObjects{
+		This:           vsanSystemEx,
+		Uuids:          []string{}, // We only need summary information.
+		Start:          0,
+		IncludeSummary: &includeSummary,
+	}
+
+	resp, err := vsanmethods.VsanQuerySyncingVsanObjects(ctx, vsanClient, &request)
+	if err != nil {
+		return err
+	}
+	fields := make(map[string]interface{})
+	fields["TotalBytesToSync"] = resp.Returnval.TotalBytesToSync
+	fields["TotalObjectsToSync"] = resp.Returnval.TotalObjectsToSync
+	fields["TotalRecoveryETA"] = resp.Returnval.TotalRecoveryETA
+	tags := populateClusterTags(make(map[string]string), clusterRef, e.URL.Host)
+	acc.AddFields(vsanSummaryMetricsName, fields, tags)
+	log.Printf("D! [inputs.vsphere][vSAN] Successfully Fetched data for Entity ==> %s:%s:%d\n", clusterRef.name, "resync", 1)
+	return nil
+}
+
+// populateClusterTags takes in a tag map, makes a copy, populates cluster related tags and returns the copy.
+func populateClusterTags(tags map[string]string, clusterRef objectRef, vcenter string) map[string]string {
+	newTags := make(map[string]string)
+	// deep copy
+	for k, v := range tags {
+		newTags[k] = v
+	}
+	newTags["vcenter"] = vcenter
+	newTags["dcname"] = clusterRef.dcname
+	newTags["clustername"] = clusterRef.name
+	newTags["moid"] = clusterRef.ref.Value
+	newTags["source"] = clusterRef.name
+	return newTags
+}
+
+// populateCMMDSTags takes in a tag map, makes a copy, adds more tags using a cmmds map and returns the copy.
+func populateCMMDSTags(tags map[string]string, entityName string, uuid string, cmmds map[string]CmmdsEntity) map[string]string {
+	newTags := make(map[string]string)
+	// deep copy
+	for k, v := range tags {
+		newTags[k] = v
+	}
+	// There are cases when the uuid is missing. (Usually happens when performance service is just enabled or disabled)
+	// We need this check to avoid index-out-of-range error
+	if uuid == "*" {
+		return newTags
+	}
+	// Add additional tags based on CMMDS data
+	if strings.Contains(entityName, "-disk") || strings.Contains(entityName, "disk-") {
+		if e, ok := cmmds[uuid]; ok {
+			if host, ok := cmmds[e.Owner]; ok {
+				if c, ok := host.Content.(map[string]interface{}); ok {
+					newTags["hostname"] = c["hostname"].(string)
```

Comment:
If these type assertions are wrong it will panic, I imagine they need to be checked.  The plugin should be able to handle bad or even malicious data without crashing.